### PR TITLE
fixed zCPU page flags and paging-related interrupts

### PR DESCRIPTION
**Pull request by Zondartul**
zCPU Documentation mentions that the memory-mapping (paging) system allows for flags to be set, which cause page access to trigger various interrupts. This behavior was previously unimplemented. This PR causes the interrupts 7, 9, 14, 28, 29, 30 to be correctly triggered by the relevant page flags upon memory page access.

### CODE REVIEW

File: lua/wire/zvm/zvm_core.lua [hunk 1]
```diff
@@ -283,22 +283,24 @@ function ZVM:Dyn_EmitInterruptCheck()
...
         self:Dyn_EmitBreak()
       self:Emit("elseif VM.MEMRQ == 3 then") -- Writing
         self:Dyn_EmitState(true)
+		self:Emit("VM.MEMRQ = 5")
         self:Emit("VM.IP = "..self.PrecompileStartIP)
         self:Emit("VM.XEIP = "..(self.PrecompileTrueXEIP or 0))
-        self:Emit("VM:Interrupt(30,VM.LADD)")
-        self:Dyn_EmitBreak()
+        self:Emit("VM:Interrupt(29,VM.LADD)")
+		self:Dyn_EmitBreak()
```
Comment (oldmud0) [inline lua/wire/zvm/zvm_core.lua]: Oops, looks like your editor changed this to tabs.

File: lua/wire/zvm/zvm_core.lua [hunk 2]
```diff
@@ -283,22 +283,24 @@ function ZVM:Dyn_EmitInterruptCheck()
...
       self:Emit("end")
     self:Emit("end")
   end
   self:Emit("if VM.INTR == 1 then")
-    self:Dyn_EmitState(true)
-    self:Dyn_EmitBreak(true)
+    --self:Dyn_EmitState(true)
```
Comment (oldmud0) [inline lua/wire/zvm/zvm_core.lua]: It's okay to just remove the old lines - if you need them back later, just look through the Git history before you made the change. That's what Git is for :wink:

File: lua/wire/zvm/zvm_features.lua [hunk 3]
```diff
@@ -283,22 +300,34 @@ function ZVM:WriteCell(Address,Value)
...
-  if self.PCAP == 1 then
+  if self.PCAP == 1 and self.MF == 1 then
     -- Fetch page
     local PageIndex = math.floor(Address / 128)
     local Page = self:GetPageByIndex(PageIndex)
 
+	if Page.Trapped == 1 then
+		self:Interrupt(30,Address) --generate interrupt and continue
+	end
+
```
Comment (oldmud0) [inline lua/wire/zvm/zvm_features.lua]: Watch for tabs: this file uses 2 spaces per indent. Also, space after `--`.

File: lua/wire/zvm/zvm_features.lua [hunk 4]
```diff
@@ -283,22 +300,34 @@ function ZVM:WriteCell(Address,Value)
...
+	--        5 - write interrupt handled
     -- Check if page is overriden
     if Page.Override == 1 then
-      if self.MEMRQ == 3 then -- Data was written
-        self.MEMRQ = 0
-        return true
+      if self.MEMRQ == 5 then -- write IRQ handled, new address/value available
+		self.MEMRQ = 0
+		Address = self.MEMADDR;
+		Value = self.LADD;
```
Comment (oldmud0) [inline lua/wire/zvm/zvm_features.lua]: Semicolons not required.

**Comment by Divran** [comment]
@PhoenixBlack

**Pull request merged**